internal/internal_workflow_client.go

```diff
@@ -333,162 +229,63 @@ func (wc *WorkflowClient) GetWorkflow(ctx context.Context, workflowID string, ru
 
 // SignalWorkflow signals a workflow in execution.
 func (wc *WorkflowClient) SignalWorkflow(ctx context.Context, workflowID string, runID string, signalName string, arg interface{}) error {
-	dataConverter := WithContext(ctx, wc.dataConverter)
-	input, err := encodeArg(dataConverter, arg)
-	if err != nil {
-		return err
-	}
-
-	request := &workflowservice.SignalWorkflowExecutionRequest{
-		Namespace: wc.namespace,
-		WorkflowExecution: &commonpb.WorkflowExecution{
-			WorkflowId: workflowID,
-			RunId:      runID,
-		},
+	return wc.interceptor.SignalWorkflow(ctx, &ClientSignalWorkflowInput{
+		WorkflowID: workflowID,
+		RunID:      runID,
 		SignalName: signalName,
-		Input:      input,
-		Identity:   wc.identity,
-	}
-
-	grpcCtx, cancel := newGRPCContext(ctx, defaultGrpcRetryParameters(ctx))
-	defer cancel()
-	_, err = wc.workflowService.SignalWorkflowExecution(grpcCtx, request)
-	return err
+		Arg:        arg,
+	})
 }
 
 // SignalWithStartWorkflow sends a signal to a running workflow.
 // If the workflow is not running or not found, it starts the workflow and then sends the signal in transaction.
 func (wc *WorkflowClient) SignalWithStartWorkflow(ctx context.Context, workflowID string, signalName string, signalArg interface{},
 	options StartWorkflowOptions, workflowFunc interface{}, workflowArgs ...interface{}) (WorkflowRun, error) {
 
-	dataConverter := WithContext(ctx, wc.dataConverter)
-	signalInput, err := encodeArg(dataConverter, signalArg)
-	if err != nil {
-		return nil, err
-	}
-
-	if workflowID == "" {
-		workflowID = uuid.NewRandom().String()
-	}
-
-	executionTimeout := options.WorkflowExecutionTimeout
-	runTimeout := options.WorkflowRunTimeout
-	taskTimeout := options.WorkflowTaskTimeout
-
-	// Validate type and its arguments.
-	workflowType, input, err := getValidatedWorkflowFunction(workflowFunc, workflowArgs, dataConverter, wc.registry)
-	if err != nil {
-		return nil, err
-	}
-
-	memo, err := getWorkflowMemo(options.Memo, wc.dataConverter)
-	if err != nil {
-		return nil, err
+	// Default workflow ID
```

Comment:
How could user not knowing the workflowID if the intention is to do SignalWithStart?

Reply:
This is existing logic. Nonetheless, if `SignalWithStart` = `Start` + `Signal`, and `Signal` uses `Start`'s workflow ID, and `Start` defaults to an auto-generated workflow ID, then one might expect `SignalWithStart` to work the way `Start` does wrt workflow ID.

I believe `SignalWithStart` can work just fine with auto-generated workflow ID just like if you manually wrote it to start then signal with the auto-generated ID from start.